openedx/core/djangoapps/content_tagging/rest_api/v1/tests/test_views.py

```diff
@@ -1044,6 +1045,121 @@ def _test_api_call(self, **kwargs) -> None:
             assert response.status_code == status.HTTP_404_NOT_FOUND
 
 
+@skip_unless_cms
+@ddt.ddt
+class TestTaxonomyUpdateOrg(TestTaxonomyObjectsMixin, APITestCase):
+    """
+    Test cases for updating orgs from taxonomies
+    """
+
+    def test_update_org(self) -> None:
+        """
+        Tests that taxonomy admin can add/remove orgs from a taxonomy
+        """
+        url = TAXONOMY_ORG_UPDATE_ORG_URL.format(pk=self.tA1.pk)
+        self.client.force_authenticate(user=self.staff)
+
+        response = self.client.put(url, {"orgs": [self.orgB.short_name, self.orgX.short_name]}, format="json")
+        assert response.status_code == status.HTTP_200_OK
+
+        # Check that the orgs were updated
+        taxonomy_orgs = TaxonomyOrg.objects.filter(taxonomy=self.tA1)
+        assert taxonomy_orgs.count() == 2
+        assert taxonomy_orgs[0].org == self.orgB
+        assert taxonomy_orgs[1].org == self.orgX
```

Comment:
For these view tests, could you instead check that the desired permissions change that was made? This lets us keep our implementation details separate from the tests.

E.g. for this test:

1. Show that user `staffA` receives a 403 when trying to update taxonomy `tA1`.
2. Update the orgs for `tA1` to include `orgA`.
3. Confirm that `staffA` can now update taxonomy `tA1` without error.

Reply:
There is no permission change here.
`tA1` = taxonomy associated with A
`staffA` = user with OrgStaffRole of orgA

From what I get from the task description, only global taxonomy admin (i.e. users with `isStaff=true`) can change taxonomy/org associations.

staffA cannot change the associations, even with tA1 associated with orgA. This check is here:
https://github.com/openedx/edx-platform/pull/33611/files#diff-297a8accac446967f48cca923cbf48ebf1f37f1b5d34235edc01d7922c588349R1140-R1160

Let me know if I got this wrong or how we can make it clear in the code/tests! :smiley:

Reply:
Ok, you raise a good point: [`tA1` is already linked to `orgA`](https://github.com/open-craft/edx-platform/blob/rpenido/fal-3542-rest-api-to-assign-a-taxonomy-to-organizations/openedx/core/djangoapps/content_tagging/rest_api/v1/tests/test_views.py#L207-L216), so `staffA` already has write access to `tA1`.

So how about this instead:

1. Login as `staffB` and show that this user receives a 403 when trying to update taxonomy`tA1`.
1. Login as `staff` to update the orgs for `tA1` to include `orgB`.
1. Login as `staffB` again and show that this user now receives a 200 when trying to update taxonomy`tA1`.

Does that make sense?

Reply:
I'm not sure if I'm following you here, but when you say update taxonomy do you mean update its metadata (name, description, etc..)?

If yes, sure! I can add a new test to explicitly check that!

Reply:
@rpenido  Sorry I wasn't clear! Yes, I meant doing a taxonomy metadata update.

Reply:
Understood! I'll update the PR tomorrow!

Reply:
Done d0fc47030d6461ef518e921dd2604ba5fa2100fb!

Reply:
Great, thank you! Can you do the same for the other (successful) org change tests?

Reply:
Done 1933c87a05e15b7e552b03bab526115098e6ca4f!

Let me know if this is not what you expected. :smiley:

Reply:
Yep, those look good.

I don't think we need tests that check that TaxonomyOrg models are changed when you call the endpoint -- that ties the implementation tightly in with the tests. But I've pestered you enough, so will let upstream/CC weigh in if they care.

Reply:
I do usually prefer that REST API test cases stick to the REST API - so if you use the POST method to update the orgs of a taxonomy, you should use the client to GET the updated taxonomy and verify that the change was made, not use the underlying Django models. In other words, I tend to agree that we shouldn't see references to the underlying `TaxonomyOrg` model in this test suite. That's an implementation detail.

Reply:
Hm.. I used to make tests more focused on one functionality than integrated tests. But I'm starting to see the benefits of what you are proposing here. I will make sure to keep that in mind for the following developments.

In this case, we don't have the GET rest API. Is it okay to leave it as it is?

Reply:
Ok! I will implement it as part of this task!

Reply:
Will we need this info in the taxonomy list and/or details? I saw something about "associated with X orgs".
If not, I think another endpoint (`GET /taxonomies/:id/orgs`) will be cleaner and avoid unnecessary data transport.

Reply:
Yes we need this on the taxonomy list, as you can see :) So it will be much more efficient to include it in the same API response.

Reply:
Done 8c5ea9994e643e6fd14a9af98c700ae211ce3b14

I think this is ready for a final review!

Reply:
Nice - thanks! I'm glad to see it was only a few extra lines of code :)